### ISSUE TITLE
Add free text boxes as alternative to required ontology fields in Experimental section

### ISSUE THREAD

**Issue by selinad** [ClinGen/clincoded#1321]
_Note:  adding this ticket now, but we need group to solidify Experimental field first -- there may be some changes to Rescue._  **please hold until this is resolved.**

We need to add free text box alternatives to several places in the Experimental section where we require ontology terms.  

Here are the places we need them.  Text boxes should be added after and _either the ontology field or the text box is required_  in all cases

1. Uberon (Expression A and Expression B) - HIGH
2. GO (Biochemical Function A and Biochemical Function B) - LOW
3. GO (Functional Alteration -- patient cells, Functional Alteration -- engineered equivalent) - HIGH
4. CL Ontology (Functional Alteration -- patient cells, Rescue patient cells) - HIGH
5. EFO (Functional Alteration -- engineered equivalent, Model Systems -- engineered equivalent, Rescue -- engineered equivalent) -- HIGH

(11 total places)

We will need to add some text to these areas as well, stating that ontology should be used whenever possible. 

Additionally, we can add some links to ontology search tools (see below:

Uberon: Search Uberon using the OLS: http://www.ebi.ac.uk/ols/ontologies/uberon
GO: Search GO using the OLS: http://www.ebi.ac.uk/ols/ontologies/go
CL: Search Cell Ontology using the OLS: http://www.ebi.ac.uk/ols/ontologies/cl
EFO: Search EFO using the OLS: http://www.ebi.ac.uk/ols/ontologies/efo

"OLS" can be linked to: http://www.ebi.ac.uk/ols/index

Additionally, we can add some pull-downs to sections that will make this easier (separate ticket).

**Comment by selinad** [comment]
We now have the information we need regarding changes to the Rescue section of experimental evidence.  I will add that as a separate ticket and link to this one.  This ticket is now ready for implementation.

**Comment by wrightmw** [comment]
Example of "GO (Biochemical Function A and Biochemical Function B)"

![screen shot 2017-04-14 at 1 59 44 pm](https://cloud.githubusercontent.com/assets/15131169/25055995/03e6fb36-211b-11e7-982a-7e38fcc88b70.png)

**Comment by wrightmw** [comment]
The text for each of the ontologies:

GO (Biochemical Function A and Biochemical Function B)
Text above text box: "If no GO term is available then describe the function of the gene below."
Text left of text box: “Identified function of gene in this record (free text):“
Error text: “Enter GO ID and/or free text“

Uberon (Expression A and Expression B)
Text above text box: "If no Uberon term is available then describe the organ type below."
Text left of text box: “Organ of tissue relevant to disease, in which gene expression is examined in patient (free text):“
Error text: “Enter Uberon ID and/or free text“

GO (Functional Alteration -- patient cells; Functional Alteration -- engineered equivalent)
NB: text box to be placed below GO ID text box.
Text above text box: "If no GO term is available then describe the normal function of the gene/gene product below."
Text left of text box: “Normal function of gene/gene product (free text):“
Error text: “Enter GO ID and/or free text“

CL Ontology (Functional Alteration -- patient cells; Rescue patient cells)
NB: text box to be placed below CL Ontology ID text box.
Text above text box: "If no CL Ontology term is available then describe the cell type below."
Text left of text box: “Patient cell type (free text):“
Error text: “Enter CL Ontology ID and/or free text“

EFO (Functional Alteration -- engineered equivalent; Rescue -- engineered equivalent)
NB: text box to be placed below EFO ID text box.
Text above text box: "If no EFO term is available then describe the cell line below."
Text left of text box: “Engineered equivalent cell type/line (free text):“
Error text: “Enter EFO ID and/or free text“

EFO (Model Systems -- engineered equivalent)
NB: text box to be placed below EFO ID text box.
Text above text box: "If no EFO term is available then describe the cell line below."
Text left of text box: “Cell culture type/line (free text):“
Error text: “Enter EFO ID and/or free text“

**Comment by selinad** [comment]
Thx @wrightmw ! Also, a reminder to include links to OLS per original ticket specification

**Comment by wrightmw** [comment]
@selinad No need to worry, @jimmyzhen is aware of the OLS links and is already adding them in.

**Comment by selinad** [comment]
Great!  Thx @wrightmw and @jimmyzhen -- double checking based on mockup.

**Comment by jimmyzhen** [comment]
Hi @selinad, @wrightmw,

I have spun up an instance for you to review the requested changes so far in this ticket:
https://1321-jz-experimental-ontologies-937553a-jzhen.demo.clinicalgenome.org

Please feel free to let me know if you have any questions. Thanks!

**Comment by selinad** [comment]
Hi @jimmyzhen -- thank you for your quick, hard work on this.  

Have just begun testing and things are looking good.  One note -- if I Edit and delete both an entered GO ID and GO free text (BF A), I see this -- I was expecting to see both fields have the error (or did something change?)

Thx!

![screen shot 2017-04-23 at 10 03 18 am](https://cloud.githubusercontent.com/assets/11320314/25315585/494b4cac-280c-11e7-84af-1353cef68a8f.png)

**Comment by selinad** [comment]
For consistency, when there are ID and free text HPO fields (e.g. BF B), have text be consistent with new text you have provided, which I prefer.  So the error text for HPO fields would be "Enter HPO ID(s) and/or free text" (current wording shown below)

![screen shot 2017-04-23 at 10 21 33 am](https://cloud.githubusercontent.com/assets/11320314/25315724/c2d63544-280e-11e7-9622-1c7d22defa01.png)

**Comment by selinad** [comment]
@jimmyzhen it looks like the GO free text field is not saved -- I had a free text desc for GO (below) -- I saved and then went to edit and it was gone:

![screen shot 2017-04-23 at 10 51 34 am](https://cloud.githubusercontent.com/assets/11320314/25315976/fe776c0e-2812-11e7-9579-bb6e7b2d3bae.png)

**Comment by selinad** [comment]
Text suggestions:

![screen shot 2017-04-23 at 11 09 09 am](https://cloud.githubusercontent.com/assets/11320314/25316082/58245774-2815-11e7-91d2-393618e49b30.png)

Suggestion for this text:
"Select a Molecular Function or Biological Process GO (Gene Ontology) term for this gene:
- View **existing GO annotations for this gene** in UniProt
- Search for a GO term using the OLS (Ontology Lookup Service)
- Search for existing or new terms using **QuickGO**"

I would be in favor of a yellow warning box above these, as we have for HPO:
![screen shot 2017-04-23 at 11 16 47 am](https://cloud.githubusercontent.com/assets/11320314/25316155/5b71d72a-2816-11e7-9580-315d32804c10.png)

The wording would change to " Please enter the gene's molecular function or biological process term (required) using the Gene Ontology (GO) term wherever possible (e.g. GO:0008150).  If no GO term exists, please email clingen-helpdesk@lists.stanford.edu and describe it in the free text box instead."

@wrightmw what do you think of asking them to email the helpdesk in all these yellow boxes?

**Comment by selinad** [comment]
Text suggestions for Expression A and B:

Yellow box:
"Please enter the relevant Uberon term for the organ of the tissue relevant to disease whenver possible (e.g. UBERON_0015228).  If no Uberon term exists, please email clingen-helpdesk@lists.stanford.edu and describe it in the free text box instead."

Above ID field:
"Search the **OLS** (Ontology Lookup Service) for a **Uberon** term"

**Comment by selinad** [comment]
Text suggestions for Functional Alteration (patient cells and engineered equivalent), in addition to GO field suggestions, above:

Yellow box:
"Please enter the relevant EFO or CLO term for the cell line/cell type whenever possible (e.g. CL_0000057, EFO_0001187). If no appropriate EFO or CLO term exists, please email clingen-helpdesk@lists.stanford.edu and describe it in the free text box instead."

Above ID field:
"Search the **OLS** (Ontology Lookup Service) for a **CLO** or **EFO** term"

(my thinking is it's best to emphasize the OLS search)

**Comment by selinad** [comment]
@jimmyzhen sorry for all the comments -- I thought it easier to separate out.  All in all, it works great!  Just the problem with the GO free text and making sure errors appear same on edit.  Thank you!

**Comment by wrightmw** [comment]
@selinad I'm happy with those text changes. The ones we had were mostly temporary placeholder texts while @jimmyzhen was implementing the free text box functionality.

**Comment by jimmyzhen** [comment]
Hi @selinad, @wrightmw,

For the GO, UBERON, CLO and EFO ID input fields, since we are already providing external linkouts in the "help text/note" above the fields, do we still need the links in the labels (see below)?
![image](https://cloud.githubusercontent.com/assets/6956310/25411516/f41ce5fc-29d1-11e7-9c15-64380f2bab93.png)

**Comment by selinad** [comment]
Good question, @jimmyzhen  -- thx.  We will be having a field for CL_ and for EFO for each of these, is that correct, @wrightmw ?  Maybe we should put "Patient cell type" and then have a CL_Ontology field and EFO field to the right? (or stacked on top of one another with an OR?)  I don't think we need the links as it's preferable they go to OLS and, as you say, the links are above...

I'll let @wrightmw comment further as he made the CL_/EFO ticket

**Comment by wrightmw** [comment]
@jimmyzhen First of all, I’d say we don’t need links in the labels.
The links are already in the text above.

**Comment by wrightmw** [comment]
@jimmyzhen If there is only one field that can be either a CLO or EFO (as you have proposed), does this confuse the data model? Would it be better to keep CLO and EFO fields separate and distinct?

**Comment by wrightmw** [comment]
Ticket #1336 is for adding multiple Uberon IDs.

**Comment by selinad** [comment]
Hi @jimmyzhen thx for your great work and for reviewing your new instance with us.

Here are the updates to the text, as we discussed -- @wrightmw, please review/edit.

### Text suggestions for Expression A and B and Functional Alteration GO sections
Yellow box
"Please enter the gene's molecular function or biological process term (required) using the Gene Ontology (GO) term wherever possible (e.g. GO:0008150). If you are unable to find an appropriate GO term, please use the free text box instead. Please email clingen-helpdesk@lists.stanford.edu for any ontology support."

### Functional Alteration AND Rescue AND Model Systems 
Yellow box -- patient cells (FA, Rescue)
"Please enter the relevant Cell Ontology (CL) term for the cell type whenever possible (e.g. CL_0000057). If you are unable to find an appropriate CL term, use the free text box instead. Please email clingen-helpdesk@lists.stanford.edu for any ontology support."

Yellow box -- engineered equivalent (FA, Rescue, Model Systems)
"Please enter the relevant EFO or Cell Ontology (CL) term for the cell line/cell type whenever possible (e.g. CL_0000057, EFO_0001187). If you are unable to find an appropriate EFO or CL term, use the free text box instead. Please email clingen-helpdesk@lists.stanford.edu for any ontology support."

Titles for fields in above sections:
- Change CLO to "Cell Ontology (CL)"
- remove EFO from title in patient cells fields for FA and Rescue and allow entry of only CL terms

### HPO yellow boxes
Change the second sentence to: "If you are unable to find an appropriate HPO term, use the free text box instead. Please email clingen-helpdesk@lists.stanford.edu for any ontology support."

### Gray writing in free text boxes associated with ontology and/or fields
For all free text boxes -- add in gray writing:
"_Use free text descriptions only after verifying no appropriate ontology term exists_"  Maybe in italics?

### Changes to Search text
"Search the OLS (Ontology Lookup Service) for a CLO or EFO term" => "Search the OLS (Ontology Lookup Service) for a Cell Ontology (CL) or EFO term"
Basically, CLO is wrong (apologies -- my bad) -- substitute "Cell Ontology (CL)" wherever it's written.  We will help check.

@wrightmw please feel free to edit.  thx @jimmyzhen !

**Comment by jimmyzhen** [comment]
Hi @selinad, @wrightmw,

Here is the latest instance for you both to review the changes:
https://1321-jz-experimental-ontologies-351952a-jzhen.demo.clinicalgenome.org

Let me know if you have any questions.

**Comment by selinad** [comment]
Hi @jimmyzhen thx for the instance!  

Looking great -- questions/comments/edits:
- can clingen-helpdesk@lists.stanford.edu open up mail? (or maybe it does, but doesn't work for me -- could be my setup)
- Use the OLS for a CL term....maybe we should have other way around?


For some reason, the rescue stuff not saving -- I had checked the boxes (see grayed entry, too) -- do I need another ticket? (this true for both Rescue choices - patient cells and engineered equivalent)
![screen shot 2017-04-27 at 4 11 15 pm](https://cloud.githubusercontent.com/assets/11320314/25508012/42288010-2b64-11e7-92d5-6ee5b53c082c.png)

**Comment by selinad** [comment]
Sorry -- I hit comment before I was really done.

Things look fantastic -- all the ontology fields behaved correctly.  I couldn't get the mail link to work, but I seem to recall having that problem before, so maybe it's me.  Everything saves.  Super great work -- thank you @jimmyzhen !

**Comment by selinad** [comment]
One more thing -- on the View pages, looks like the terms are going to Ontobee -- can they go to the OLS entry without too much trouble (we could do for R12)

**Comment by selinad** [comment]
For some reason, the free text field not tabbed over on this one (for patient cell type)
![screen shot 2017-04-27 at 4 19 41 pm](https://cloud.githubusercontent.com/assets/11320314/25508211/60f8b1d0-2b65-11e7-833a-020207e5a652.png)

**Comment by jimmyzhen** [comment]
Hi @selinad,

Great catch on the mail link problem! I found the bug and will fix that. Thanks!

**Comment by jimmyzhen** [comment]
Hi @selinad, @wrightmw,

I have spun up a new instance for you to review the changes in addressing the reported issues (see above):
https://1321-jz-experimental-ontologies-36932c0-jzhen.demo.clinicalgenome.org

Please let me know if there are any questions. Thanks.

**Comment by jimmyzhen** [comment]
Hi @selinad,

Thank you for explaining how you encountered that checkbox bug for the "Rescue" experimental type.

I have fixed the bug and you may now verify the fix on the very same instance.

I also checked the other experimental types and their checkboxes appeared to be working as expected. Please let me know if you have any questions.

**Comment by wrightmw** [comment]
@jimmyzhen 
BiochemA and B:
Minor text change: "Search for a GO using the OLS (Ontology Lookup Service)." should say "Search for a GO term using the OLS (Ontology Lookup Service)."

**Comment by selinad** [comment]
Hi @jimmyzhen -- looking great so far!  

On the view page for Expression A, this Uberon ID goes to an OLS 404
![screen shot 2017-05-04 at 4 22 34 pm](https://cloud.githubusercontent.com/assets/11320314/25728778/0bc23944-30e6-11e7-8151-0ab3549b20f4.png)
![screen shot 2017-05-04 at 4 23 24 pm](https://cloud.githubusercontent.com/assets/11320314/25728779/0dc28de8-30e6-11e7-89be-1928de94f215.png)

**Comment by selinad** [comment]
Hi @jimmyzhen 
for GO term links on the View page, I think we could go to the OLS page as well instead of QuickGO, if it's a simple change.  QuickGO is slow and unwieldy.

Here's an example: http://www.ebi.ac.uk/ols/ontologies/go/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FGO_2001284

Also, GO:2001284 might be nice to use as an example ID -- the one you have now (GO:0008150) goes to the very top GO term, which is meaningless (sorry I haven't caught that before)

**Comment by selinad** [comment]
@jimmyzhen everything is looking great!  All functionalities behaving with check boxes, view pages good, text good except @wrightmw 's good catch above -- thank you!

**Comment by jimmyzhen** [comment]
Hi @selinad.

Would you also want to change the QuickGO links to the OLS page on the experimental data curation form page as well (see screenshot below)?
![image](https://cloud.githubusercontent.com/assets/6956310/25816004/8ae1ed70-33d7-11e7-9214-692fb155919f.png)

**Comment by jimmyzhen** [comment]
Hi @selinad, @wrightmw,

I have addressed change requests and you may now verify them again on the very same instance:
https://1321-jz-experimental-ontologies-36932c0-jzhen.demo.clinicalgenome.org

**Comment by selinad** [comment]
Hi @jimmyzhen -- re QuickGO links on the experimental data -- I think it's OK there as it's a 3rd option for them....

**Comment by jimmyzhen** [comment]
Great. Thanks, Selina!

**Comment by selinad** [comment]
Hi @jimmyzhen -- everything looks fantastic!
- text change caught by @wrightmw is fixed
- GO links from View page look great
- new example GO term seems to be in place for all evidence types

Let me know if I missed testing anything.

Great work -- thx @jimmyzhen !

**Comment by wrightmw** [comment]
@jimmyzhen I agree. Great work!

**Comment by kilodalton** [comment]
Included in last release (R11). Nice job and thanks for your hard work.
